mrmustard/lab_dev/sampler.py

```diff
@@ -0,0 +1,192 @@
+# Copyright 2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Samplers for measurement devices.
+"""
+
+from __future__ import annotations
+from itertools import product
+
+from abc import ABC, abstractmethod
+
+from typing import Any, Sequence
+
+import numpy as np
+
+from mrmustard import math, settings
+
+from .states import State, Number, QuadratureEigenstate
+from .circuit_components import CircuitComponent
+from .circuit_components_utils import BtoQ
+
+__all__ = ["Sampler", "PNRSampler", "HomodyneSampler"]
+
+
+class Sampler(ABC):
+    r"""
+    A sampler for measurements of quantum circuits.
+
+    Args:
+        meas_outcomes: The measurement outcomes for this sampler.
+        povms: The POVMs of this sampler.
+    """
+
+    def __init__(
+        self,
+        meas_outcomes: Sequence[Any],
+        povms: CircuitComponent | Sequence[CircuitComponent],
+    ):
```

Comment:
Perhaps we should check that `len(meas_outcomes)` is the same as `len(povms)` (or `len(povms) - 1` in case we let the Sampler complete the POVM).